Fix failing to run when launched with npx

Codex- opened the pull request:
When testing this before adding this to the package scripts I found I was unable to run `cyclonedx-npm`, due to the command it was selecting being an invalid command to spawn a process with:
```
 >  npx cyclonedx-npm
DEBUG | options: {"packageLockOnly":false,"omit":[],"flattenComponents":false,"specVersion":"1.4","outputFormat":"JSON","outputFile":"-","mcType":"application"}
DEBUG | packageFile: /Users/alexmiller/dev/app/package.json
DEBUG | projectDir: /Users/alexmiller/dev/app
DEBUG | lockFile: /Users/alexmiller/dev/app/package-lock.json
INFO  | gather dependency tree ...
DEBUG | npm-ls: run /Users/alexmiller/.asdf/installs/nodejs/16.14.2/.npm/lib/node_modules/npm/bin/npx-cli.js with ["ls","--json","--all","--long"] in /Users/alexmiller/dev/app
WARN  | npm-ls: STDERR
  npm ERR! could not determine executable to run
  
  npm ERR! A complete log of this run can be found in:
  npm ERR!     /Users/alexmiller/.npm/_logs/2022-08-30T23_38_21_054Z-debug-0.log
  
ERROR | npm-ls: errors
  {}
/Users/alexmiller/dev/app/node_modules/@cyclonedx/cyclonedx-npm/dist/builders.js:82
            throw new Error(`npm-ls exited with errors: ${error.errno ?? '???'} ${error.code ?? npmLsReturns.status ?? 'noCode'} ${error.signal ?? npmLsReturns.signal ?? 'noSignal'}`);
            ^

Error: npm-ls exited with errors: ??? 1 noSignal
    at BomBuilder.fetchNpmLs (/Users/alexmiller/dev/app/node_modules/@cyclonedx/cyclonedx-npm/dist/builders.js:82:19)
    at BomBuilder.buildFromLockFile (/Users/alexmiller/dev/app/node_modules/@cyclonedx/cyclonedx-npm/dist/builders.js:43:41)
    at Object.run (/Users/alexmiller/dev/app/node_modules/@cyclonedx/cyclonedx-npm/dist/cli.js:97:19)
    at Object.<anonymous> (/Users/alexmiller/dev/app/node_modules/@cyclonedx/cyclonedx-npm/bin/cyclonedx-npm-cli.js:2:27)
    at Module._compile (node:internal/modules/cjs/loader:1103:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1157:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
    at node:internal/main/run_main_module:17:47
```

We can see from this that it's trying to use: `/Users/alexmiller/.asdf/installs/nodejs/16.14.2/.npm/lib/node_modules/npm/bin/npx-cli.js` as provided by using through `npx`. However, this strictly requires the use of `npm`

I have added some simple handling of the case where `npx` is used to run `cyclonedx-npm`.

Running:
```
> npx cyclonedx-npm
DEBUG | options: {"packageLockOnly":false,"omit":[],"flattenComponents":false,"specVersion":"1.4","outputFormat":"JSON","outputFile":"-","mcType":"application"}
DEBUG | packageFile: /Users/alexmiller/dev/app/package.json
DEBUG | projectDir: /Users/alexmiller/dev/app
DEBUG | lockFile: /Users/alexmiller/dev/app/package-lock.json
DEBUG | command: npx-cli.js usage detected, checking for npm-cli.js ...
DEBUG | command: npm-cli.js found
DEBUG | command: /Users/alexmiller/.asdf/installs/nodejs/16.14.2/.npm/lib/node_modules/npm/bin/npm-cli.js
INFO  | gather dependency tree ...
DEBUG | npm-ls: run /Users/alexmiller/.asdf/installs/nodejs/16.14.2/.npm/lib/node_modules/npm/bin/npm-cli.js with ["ls","--json","--all","--long"] in /Users/alexmiller/dev/app
```

jkowalleck commented:
i see you want to enable the tool to run via `npx` instead of `npm exec`.
Could you open an issue describing the intended behavior, so that a regression test can be shaped?

jkowalleck commented on src/builders.ts:
```diff
@@ -90,11 +91,33 @@ export class BomBuilder {
+    if (/npx-cli.js/gi.test(command)) {
+      // `npm` must be used for executing `ls`.
+      // Typically `npm-cli.js` is alongside `npx-cli.js`, as such we attempt to use this and validate it exists.
+      const npmPath = command.replace('npx-cli.js', 'npm-cli.js')
```
:x: replace the trailing "npx-cli.js" via `/npx-cli\.js$/

attention: your initial regex-test used case-insensitivity (`i`) -- it might need to apply here, again.

Codex- replied:
shouldn't be insensitive, but also the script filename is case sensitive so they won't randomly make it `NpX-cLi.js` any time soon. 

Removed the `i` flag.

jkowalleck commented on src/builders.ts:
```diff
@@ -90,11 +91,33 @@ export class BomBuilder {
-    const command = process.env.npm_execpath || 'npm'
+    let command = process.env.npm_execpath || 'npm'
+
+    if (/npx-cli.js/gi.test(command)) {
```
was this tested on windows, too?

jkowalleck commented on src/builders.ts:
```diff
@@ -90,11 +91,33 @@ export class BomBuilder {
+    }
 
+    this.console.debug(`DEBUG | command: ${command}`)
+    return command
```
:x: dont store the value in a variable to have it eventually returned.

just store a const and return it as a fallaback.
all those modifications to the variable could be transformed to `return` easily,

example (simplified, not actual code):
```javascript
private getNpmCommand (process: NodeJS.Process): string {
  const npm_execpath = process.env.npm_execpath ?? ''

  if ( /npx-foo/.test(npm_execpath)  ) {
    const npm = npm_execpath.replace('npx','npm')
    return fs.existsSycn(npm)
      ? npm  // detected npm
      : 'npm' // fallback to global npm
  }

  return npm_execpath || 'npm'
}

```

jkowalleck reviewed (changes requested):
thank you for your contribution.

I marked critical points with an :x:  - these should be change.
other remarks are just notes or questions.

Codex- commented on src/builders.ts:
```diff
@@ -90,11 +91,33 @@ export class BomBuilder {
-    const command = process.env.npm_execpath || 'npm'
+    let command = process.env.npm_execpath || 'npm'
+
+    if (/npx-cli.js/gi.test(command)) {
```
Sure, the leading `/` was omitted for path compatibility but can make it explicit.

Codex- commented:
> i see you want to enable the tool to run via `npx` instead of `npm exec`. Could you open an issue describing the intended behavior, so that a regression test can be shaped?

Currently, this application can only work if you install it as a local dependency, globally installing it does not work and it doesn't make a lot of sense in many cases to have this as a dependency when it can be simply tied into CICD.

jkowalleck commented:
re: https://github.com/CycloneDX/cyclonedx-node-npm/pull/91#issuecomment-1237493834

have you tried `npm exec --` instead of `npx` ? 
see https://github.com/CycloneDX/cyclonedx-node-npm/pull/105/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R56-R59 

your feedback is appreciated

Codex- commented:
> have you tried `npm exec --` instead of `npx` ? see https://github.com/CycloneDX/cyclonedx-node-npm/pull/105/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R56-R59

`npm exec` works of course, because it has the compatible `npm_execpath` in the env. 

`npx` usage is very common

jkowalleck commented:
#105 got merged.

please rebase onto latest `1.0-dev` branch, 
and update the README accordingly, and refere to `npx`, where it currently describes `npm exec`

jkowalleck commented on src/builders.ts:
```diff
@@ -90,11 +107,32 @@ export class BomBuilder {
+    // This can be `npm`, `npx`, or `undefined` if called by `node` directly.
     /* eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing, @typescript-eslint/strict-boolean-expressions -- need to handle optional empty-string */
-    const command = process.env.npm_execpath || 'npm'
+    const execPath = process.env.npm_execpath || ''
```
:x: use `??` here

Codex- commented on src/builders.ts:
```diff
@@ -59,6 +60,22 @@ export class BomBuilder {
+   *   - foo/npx-cli_js     // Invalid extension
+   *   - npx-cli.js/foo.sh  // Directory of the same name
+   */
+  private readonly npxMatcher = new RegExp(`(?:^|${sep})npx-cli\\.js$`, 'g')
```
Sounds good, I have to get my windows VM running again to test it again which i'd want to do before this is merged anyway.

jkowalleck replied:
:x: yupp. crashed on windows: 
https://github.com/CycloneDX/cyclonedx-node-npm/runs/8197068636?check_suite_focus=true

jkowalleck replied:
maybe just add ` \\` inforont of the `${sep}` so it gets regex-escaped properly ...

Codex- replied:
Cool, i'll get on it. I'll move this into draft until I can retest it locally on windows.

jkowalleck replied:
no worries, we have a CI/CT pipeline to run windows test for you :-)

Codex- commented on src/builders.ts:
```diff
@@ -90,11 +107,32 @@ export class BomBuilder {
+      this.console.debug('DEBUG | command: npx-cli.js usage detected, checking for npm-cli.js ...')
+      // Typically `npm-cli.js` is alongside `npx-cli.js`, as such we attempt to use this and validate it exists.
+      // Replace the script in the path, and normalise it with resolve (eliminates any extraneous path separators).
+      const npmPath = resolve(execPath.replace(this.npxMatcher, `${sep}npm-cli.js`))
```
Fair point, will swap

jkowalleck commented:
pullrequest looks promissing, all tests are passing. Thank you very much.

please add yourself to the contributors list in the `package.json` file,
and write an appropriate entry in the `HISTORY.md` file

jkowalleck commented on src/builders.ts:
```diff
@@ -90,11 +107,29 @@ export class BomBuilder {
+        : 'npm' // fallback to global npm
+    }
+
+    return execPath ?? 'npm'
```
:x: this must be `execPath || 'npm'` still, so empty strings get taken care of.

Codex- commented:
> pullrequest looks promising, all tests are passing. Thank you very much.
> 
> feel free add yourself to the contributors list in the `package.json` file - if you like, and please write an appropriate entry in the `HISTORY.md` file.

Awesome, thank you. 

Fixed one last issue where the path replacement wasn't working as expected (due to `?:` specifying a non-capture group).

Rebased on latest.
